File: samples/hello_grisp/src/hello.erl
-module(hello).

-on_load(init/0).

-export([init/0, world/0]).

-define(APPNAME, hello_grisp).
-define(LIBNAME, hello_grisp).

world() ->
      "NIF library not loaded".

init() ->
    SoName = case code:priv_dir(?APPNAME) of
        {error, bad_name} ->
            case filelib:is_dir(filename:join(["..", priv])) of
                true -> filename:join(["..", priv, ?LIBNAME]);
                _ -> filename:join([priv, ?LIBNAME])
            end;
        Dir -> filename:join(Dir, ?LIBNAME)
    end,
    erlang:load_nif(SoName, 0).

File: samples/hello_elixir/lib/hello.ex
defmodule Hello do
  @moduledoc false

  @on_load :load_nif

  def load_nif do
    case :code.priv_dir(:hello_elixir) do
      dir when is_list(dir) ->
        path = :filename.join(dir, ~c"hello")
        case :erlang.load_nif(path, 0) do
          :ok -> :ok
          {:error, _} -> :ok
        end

      _ ->
        :ok
    end
  end

  def world, do: :erlang.nif_error(:nif_not_loaded)
end
